workshop/exercise-2.md

```diff
@@ -0,0 +1,318 @@
+
+Crea un nuevo paquete `config` y dentro una clase `ChatConfig`.
+
+```java
+    // src/main/java/com/theitdojo/optimizing_llm_responses_with_rag_in_java/config/ChatConfig.java
+    package com.theitdojo.optimizing_llm_responses_with_rag_in_java.config;
+
+    import org.springframework.ai.chat.memory.ChatMemory;
+    import org.springframework.ai.chat.memory.InMemoryChatMemory;
+    import org.springframework.context.annotation.Bean;
+    import org.springframework.context.annotation.Configuration;
+
+    @Configuration
+    public class ChatConfig {
+
+        @Bean
+        public ChatMemory chatMemory() {
+            return new InMemoryChatMemory();
+        }
+    }
+```
+
+3. Usar la Memoria en las Llamadas al LLM
+
+Ahora, actualiza los métodos de `ChatAssistantService` para que utilicen el `conversationId` y la `chatMemory` en cada llamada. 
+
+Esto se hace encadenando `.chatReference(conversationId)` y `.chatMemory(chatMemory)` en la llamada al `prompt`. No olvides actualizar las firmas de los métodos para que coincidan con la interfaz.
+
+```java
+    // Dentro de ChatAssistantService.java
+
+    @Override
+    public String getResponse(String conversationId, String message) {
+        return this.chatClient.prompt()
+                .chatReference(conversationId) // Identifica la conversación
+                .chatMemory(this.chatMemory)   // Usa la memoria
+                .user(message)
+                .call()
+                .content();
+    }
+
+    @Override
+    public Stream<String> streamResponse(String conversationId, String message) {
+        return chatClient.prompt()
+                .chatReference(conversationId) // Identifica la conversación
+                .chatMemory(this.chatMemory)   // Usa la memoria
+                .user(message)
+                .stream()
+                .content()
+                .toStream();
+    }
+    
+    // ... Asegúrate de actualizar también askQuestion y askQuestionWithContext
+```
+
+## Hora de probar la memoria del asistente (In-Memory)
+
+Para apreciar realmente el valor de la persistencia, vamos a realizar pruebas en dos fases.
+
+### Fase 1: Probando la Memoria Volátil (con InMemoryChatMemory)
+
+Antes de cambiar a la memoria persistente, primero vamos a verificar el comportamiento de la implementación en memoria que configuraste en la Parte 2.
+
+1. Asegúrate de que tu `ChatConfig` esté proveyendo el bean de `InMemoryChatMemory`. Si ya lo cambiaste, vuelve a ponerlo temporalmente.
+
+```java
+// En ChatConfig.java
+@Bean
+public ChatMemory chatMemory() {
+    return new InMemoryChatMemory();
+}
+```
+
+2. Ejecuta la aplicación.
+3. Abre tu navegador y haz una pregunta para establecer contexto: `http://localhost:8080/ai/chat?message=Mi nombre es Fulano`.
+3. Haz una pregunta inicial: `http://localhost:8080/ai/chat?message=Mi nombre es Fulano. ¿Puedes decirme qué es Spring AI`.
```

Comment:
Should we only keep the question here? Given that we sent the name in the previous message
```suggestion
3. Haz una pregunta inicial: `http://localhost:8080/ai/chat?message=¿Puedes decirme qué es Spring AI?`.
```